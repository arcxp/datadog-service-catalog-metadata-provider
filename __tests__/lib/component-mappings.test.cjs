const YAML = require('yaml')
const core = require('@actions/core')
const subject = require('../../lib/component-mappings.cjs')
const { readFileSync } = require('node:fs')

// set up the schema validators
const Ajv = require('ajv')
const schema = new Ajv({ strict: false, validateFormats: false })
schema.addSchema(require('../data/v3/metadata.schema.json'), 'metadata')
schema.addSchema(require('../data/v3/entity.schema.json'), 'entity')
schema.addSchema(require('../data/v3/integration.schema.json'), 'integration')

// Adding refs
schema.addSchema(
  require('../data/v3/integration_opsgenie.schema.json'),
  'integration_opsgenie.schema.json'
)
schema.addSchema(
  require('../data/v3/integration_pagerduty.schema.json'),
  'integration_pagerduty.schema.json'
)

describe('imports', () => {
  test('verify imports', () => {
    expect(subject).toMatchObject({
      mapComponents: expect.any(Function),
      _test: expect.objectContaining({
        mapComponentMetadata: expect.any(Function),
        mapEntityComponent: expect.any(Function),
        mapServiceComponent: expect.any(Function),
        mapDatastoreComponent: expect.any(Function),
      }),
    })
  })
})

// Test the various schema individually and collectively
describe.each([
  {
    label: 'metadata',
    func: subject._test.mapComponentMetadata,
    funcName: 'mapComponentMetadata',
    sample: 'v3-metadata.yaml',
  },
  {
    label: 'entity',
    func: subject._test.mapEntityComponent,
    funcName: 'mapEntityComponent',
    sample: 'v3-entity.yaml',
  },
  {
    label: 'integration',
    func: subject._test.mapIntegration,
    funcName: 'mapIntegration',
    sample: 'v3-integrations.yaml',
  },
])('$label - $funcName() - $sample', ({ label, func, funcName, sample }) => {
  const validate = schema.getSchema(label)

  // Let's test the entity types
  test(`#${funcName}() - empty`, () => {
    // Empty works
    expect(func()).toBeUndefined()
  })

  test(`#${funcName}() - full`, () => {
    const parsedTestInput = YAML.parse(
      readFileSync(`__tests__/data/${sample}`, {
        encoding: 'utf8',
      })
    )
    const output = func(core, parsedTestInput)
    expect(output).toMatchSnapshot()
    const doesItValidate = validate(output)
    if (validate.errors) {
      console.debug(validate.errors)
    }
    expect(doesItValidate).toStrictEqual(true)
    expect(validate.errors).toBeNull()
  })
})
