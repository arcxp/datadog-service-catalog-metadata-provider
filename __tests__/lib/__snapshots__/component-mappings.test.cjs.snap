// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity - mapEntityComponent() - v3-entity.yaml #mapEntityComponent() - full 1`] = `
{
  "apiVersion": "v3",
  "integrations": {
    "opsgenie": {
      "region": "US",
      "serviceURL": "https://opsgenie.example.com/service-url",
    },
    "pagerduty": {
      "serviceURL": "https://pagerduty.example.com/service-url",
    },
  },
  "kind": "system",
  "metadata": {
    "additionalOwners": [
      {
        "name": "sre-team",
        "type": "operator",
      },
    ],
    "contacts": [
      {
        "contact": "https://my-org.slack.com/archives/oncall",
        "name": "Oncall Slack",
        "type": "slack",
      },
    ],
    "description": "A service for handling shopping cart interactions",
    "displayName": "Shopping Cart Service",
    "links": [
      {
        "name": "Runbook",
        "provider": "Confluence",
        "type": "runbook",
        "url": "https://my-runbook.com",
      },
    ],
    "name": "shopping-cart",
    "owner": "shopping-team",
    "tags": [
      "e-commerce",
      "cart",
    ],
  },
}
`;

exports[`integration - mapIntegration() - v3-integrations.yaml #mapIntegration() - full 1`] = `
{
  "opsgenie": {
    "region": "US",
    "serviceURL": "https://opsgenie.example.com/service-url",
  },
  "pagerduty": {
    "serviceURL": "https://pagerduty.example.com/service-url",
  },
}
`;

exports[`metadata - mapComponentMetadata() - v3-metadata.yaml #mapComponentMetadata() - full 1`] = `
{
  "additionalOwners": [
    {
      "name": "additionalTeam1",
      "type": "support",
    },
    {
      "name": "additionalTeam2",
      "type": "qa",
    },
  ],
  "contacts": [
    {
      "contact": "https://my-org.slack.com/archives/my-channel",
      "name": "Oncall Slack",
      "type": "slack",
    },
    {
      "contact": "contact@datadoghq.com",
      "name": "Team Email",
      "type": "email",
    },
  ],
  "description": "Short description of my service.",
  "displayName": "My Service",
  "inheritFrom": "application:default/myapp",
  "links": [
    {
      "name": "Runbook",
      "provider": "Confluence",
      "type": "runbook",
      "url": "https://my-runbook.com",
    },
    {
      "name": "Dashboard",
      "provider": "Datadog",
      "type": "dashboard",
      "url": "https://my-dashboard.com",
    },
  ],
  "name": "myService",
  "owner": "myTeam",
  "tags": [
    "my:tag",
    "environment:production",
  ],
}
`;
